### PR TITLE
Fix Expiry date issues

### LINKED ISSUE #161: [PE-D][Tester E] Can add expired medication
Allows me to add medication with expiry date that is not a date. Is this intended?

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: itsmejr257/ped#8

### LINKED ISSUE #186: [PE-D][Tester F] adding a non-existant date is allowed
-------------
Labels: `severity.VeryLow` `type.FunctionalityBug`
original: nikhil-2101/ped#2

### LINKED ISSUE #187: [PE-D][Tester D] Check for valid expiry date
`  4. name | 1000.0 | 11/40/2031 | null`
For now I can also add in date that does not make sense.

-------------
Labels: `severity.Low` `type.FeatureFlaw`
original: fxe025/ped#6

### LINKED ISSUE #188: [PE-D][Tester F] using a different date format while not following the  correct format of adding medication is allowed
-------------
Labels: `type.FunctionalityBug` `severity.Medium`
original: nikhil-2101/ped#3

### LINKED ISSUE #192: [PE-D][Tester F] another form of date format issue is present
-------------
Labels: `severity.Low` `type.FunctionalityBug`
original: nikhil-2101/ped#6

### LINKED ISSUE #202: [PE-D][Tester D] Utilise the expiry date feature
Implement some methods to check if the medicine is expired from the expiry date and current time. Taking expired medicine is unsafe and they should be removed from list of medicine to be taken.

-------------
Labels: `severity.Medium` `type.FeatureFlaw`
original: fxe025/ped#13

### PR DESCRIPTION
**Pull request by e0958902**
# Checklist for feature implementation and modification
Fixes #202, Fixes #192, Fixes #188, Fixes #187, Fixes #186, Fixes #161
- [x] Added relevant milestone (under milestone tab on the right)
- Implementation
  - [ ] Sources (if any) credited under both relevant section(s) and Developer Guide (Acknowledgement)
  - [x] JUnit Test(s)
- Documentation
  - [ ] JavaDoc for classes and methods
  - Developer Guide
    - [ ] Design (Can add as a todo statement for later on)
    - [ ] Implementation (Can add as a todo statement for later on)
    - [ ] Relevant User Story present
    - [ ] Non-functional requirement(s) that the feature addresses (if any)
    - [ ] Glossary of the (technical) term(s) used (if any)
    - [ ] Specialised instructions for manual testing (if any)
  - User Guide
    - [ ] Description and usage of feature (if any)
  - Contributions (under `/docs/team/*.md`)
    - [ ] Code/Feature/Enhancement contribution
    - [ ] Developer Guide contribution
    - [ ] User Guide contribution
    - [ ] Misc and team-based admin tasks (if any)

# Checklist for non-implementation related PR
Fixes #202, Fixes #192, Fixes #188, Fixes #187, Fixes #186, Fixes #161
- [ ] Added relevant milestone (under milestone tab on the right)
- Documentation (Contributions)
  - [ ] Developer Guide contributions (if any)
  - [ ] User Guide contributions (if any)
  - [ ] Review contributions (if any)
  - [ ] Miscellaneous and team-based admin tasks (if any)

# Additional Information
Nil

### CODE REVIEW

File: src/main/java/meditracker/medication/MedicationManager.java [hunk 1]
```diff
@@ -72,6 +77,30 @@ private static void checkForDuplicateMedication(String name) throws MediTrackerE
...
 
+    /**
+     * Checks if the user has input a valid expiry date format.
+     * Also checks that the expiry date entered is not expired.
+     *
+     * @param expiryDate Expiry Date in yyyy-MM-dd format
+     * @throws DateTimeParseException When the date entered is in the wrong format
+     * @throws MediTrackerException When the date entered is already expired
+     */
+    public static void checkforValidExpiryDate(String expiryDate)
```
Comment (wenenhoe) [inline src/main/java/meditracker/medication/MedicationManager.java:88]: Might be better to have this method in `Medication` instead as it does not have to deal with the context of all `Medication` but just itself only

File: src/main/java/meditracker/medication/MedicationManager.java [hunk 2]
```diff
@@ -72,6 +77,30 @@ private static void checkForDuplicateMedication(String name) throws MediTrackerE
...
+            LocalDate parsedExpiryDate = LocalDate.parse(expiryDate, dateTimeFormatter);
+            LocalDate currentDate = MediTrackerTime.getCurrentDate();
+
+            if (parsedExpiryDate.isBefore(currentDate)) {
+                throw new MediTrackerException("You are not allowed to enter expired medications!");
+            }
+
+        } catch (DateTimeParseException e) {
+            throw new MediTrackerException("Please enter a valid expiry date in yyyy-MM-dd!");
+        }
```
Comment (wenenhoe) [inline src/main/java/meditracker/medication/MedicationManager.java:101]: Might be better to scope the try catch to the line(s) that will throw the exception

```suggestion
        DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
        LocalDate parsedExpiryDate;
        try {
            parsedExpiryDate = LocalDate.parse(expiryDate, dateTimeFormatter);
        } catch (DateTimeParseException e) {
            throw new MediTrackerException("Please enter a valid expiry date in yyyy-MM-dd!");
        }
        LocalDate currentDate = MediTrackerTime.getCurrentDate();
        if (parsedExpiryDate.isBefore(currentDate)) {
            throw new MediTrackerException("You are not allowed to enter expired medications!");
        }
```

File: src/main/java/meditracker/medication/MedicationManager.java [hunk 3]
```diff
@@ -72,6 +77,30 @@ private static void checkForDuplicateMedication(String name) throws MediTrackerE
...
+    /**
+     * Checks if the user has input a valid expiry date format.
+     * Also checks that the expiry date entered is not expired.
+     *
+     * @param expiryDate Expiry Date in yyyy-MM-dd format
+     * @throws DateTimeParseException When the date entered is in the wrong format
+     * @throws MediTrackerException When the date entered is already expired
+     */
+    public static void checkforValidExpiryDate(String expiryDate)
+            throws DateTimeParseException, MediTrackerException {
```
Comment (wenenhoe) [inline src/main/java/meditracker/medication/MedicationManager.java:89]: `DateTimeParseException` is not being thrown by this method but rather caught and handled with, do remove it
Likewise, remove it from the `addMedication` method too

**Review by wenenhoe — CHANGES_REQUESTED** [review]
Thanks for implementing the expiry date logic. Requested for some changes to be made, do take a look

File: src/main/java/meditracker/medication/MedicationManager.java (hunk 1 above)
Comment (e0958902) [reply src/main/java/meditracker/medication/MedicationManager.java]: Implemented in [de50cef](https://github.com/AY2324S2-CS2113T-T09-1/tp/pull/220/commits/de50cefd7e288993053326b5d43ea2fce6ef35d3)

File: src/main/java/meditracker/medication/MedicationManager.java (hunk 2 above)
Comment (e0958902) [reply src/main/java/meditracker/medication/MedicationManager.java]: Implemented in [03dcdc0](https://github.com/AY2324S2-CS2113T-T09-1/tp/pull/220/commits/03dcdc0cbbfea652ba737bdb9ce7952053dd7d18)

File: src/main/java/meditracker/medication/MedicationManager.java (hunk 3 above)
Comment (e0958902) [reply src/main/java/meditracker/medication/MedicationManager.java]: Implemented in [41b4727](https://github.com/AY2324S2-CS2113T-T09-1/tp/pull/220/commits/41b4727d6e9034177cc15b2e64c72417d26cf55e)

**Review by wenenhoe — APPROVED** [review]
LGTM! Thanks for making the changes

**Pull request merged**